File: gcpc2016/correctingcheeseburgers/submissions/time_limit_exceeded/StefanStringBFS.java
//Author: Stefan Toman

import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class StefanStringBFS {
  public static void main(String[] args) {
    //read input
    Scanner s = new Scanner(System.in);
    int n = s.nextInt(), t;
    String start = "", end = "";
    for(int i = 0; i < n; i++) {
      t = s.nextInt() - 1;
      start += t;
      end += i;
    }
    Set<String> reachable_start = new HashSet<String>();
    reachable_start.add(start);                

    //compute reachable nodes
    t = 0;
    while(!reachable_start.contains(end)) {
      reachable_start = step(n, reachable_start, true);
      t++;
    }

    System.out.println(t);
  }
  
  public static Set<String> step(int n, Set<String> reachable, boolean forward) {
    Set<String> ret = new HashSet<String>();
    for(String t: reachable) {
      for(int b = 0; b <= n; b++) {
        for(int c = b; c <= n; c++) {
          for(int d = c; d <= n; d++) {
            String pa = t.substring(0, b);
            String pb = t.substring(b, c);
            String pc = t.substring(c, d);
            String pd = t.substring(d);
            String p = forward ? pc + pa + pd + pb : pb + pd + pa + pc;
            if(!ret.contains(p)) {
              ret.add(p);
            }
          }
        }
      }                                                                                                                                                
    }
    
    return ret;
  }
}

File: gcpc2016/correctingcheeseburgers/submissions/time_limit_exceeded/stefan_string_bfs.cpp
//Author: Stefan Toman

#include <iostream>
#include <unordered_set>
#include <string>

using namespace std;

//get all bit-shuffles of a set
unordered_set<string> step(int n, unordered_set<string> reachable, bool forward) {
  unordered_set<string> ret;
  for(auto it = reachable.begin(); it != reachable.end(); ++it) {
    for(int b = 0; b <= n; b++) {
      for(int c = b; c <= n; c++) {
        for(int d = c; d <= n; d++) {
          string pa = it->substr(0, b);
          string pb = it->substr(b, c-b);
          string pc = it->substr(c, d-c);
          string pd = it->substr(d, n-d);
          string p = forward ? pc + pa + pd + pb : pb + pd + pa + pc;
          if(ret.find(p) == ret.end()) {
            ret.insert(p);
          }
        }
      }
    }
  }

  return ret;
}

int main() {
  //read input
  int n, t;
  string start = "", end = "";
  cin >> n;
  for(int i = 0; i < n; i++) {
    cin >> t;
    start += to_string(t-1);
    end += to_string(i);
  }
  unordered_set<string> reachable_start ({start});
  
  //compute reachable nodes
  t = 0;
  while(reachable_start.find(end) == reachable_start.end()) {
    reachable_start = step(n, reachable_start, true);
    t++;
  }
  
  cout << t << endl;
  return 0;
}
